### PR TITLE
Add SNI configuration docs

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/26242

@halter73 Is there important information I missed? On the surface it seems quite simple to explain (that's good)

Internal preview: https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0&branch=pr-en-us-21941

### CODE REVIEW

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 1]
```diff
@@ -314,7 +321,57 @@ webBuilder.ConfigureKestrel(serverOptions =>
...
+    }
+  }
+}
+```
+
+The server name supports wildcard matching:
+
+* Exact match. `a.example.org` matches `a.example.org`.
+* Wildcard prefix. `*.example.org` matches `b.example.org` and `c.example.org`.
+* Full wildcard. `*` matches everything else.
```
[the remark this answers is not in the document]
Comment (halter73) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:372]: > If no relevant configuration exists in the selected SNI or named endpoint section for a given property, and no defaults are configured via "EndpointDefaults" and/or "Certificates:Default/Development", configuration will fallback to what was configured via KSO.ConfigureHttpsDefaults(Action<HttpsConnectionAdapterOptions> configureOptions) or KSO.ConfigureEndpointDefaults(Action<ListenOptions> configureOptions).

Should we mention anything about this or is this unsurprising enough to be left unstated?

[the remark this answers is not in the document]
Comment (JamesNK) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:372]: > I would add that Kestrel will refuse the connection if there's matching pattern.

Pardon?

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 2]
```diff
@@ -314,7 +321,57 @@ webBuilder.ConfigureKestrel(serverOptions =>
...
+              "Path": "<path to .pfx file>",
+              "Password": "<certificate password>"
+            }
+          },
+          "*": {
+            // At least one subproperty needs to exist per SNI section or it
+            // cannot be discovered via IConfiguration
+            "Protocols": "Http1",
+          }
+        }
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:355]: I've mentioned SNI configuration is applied to an endpoint in the description. Is it clear enough?

I'll leave JSON as is. I don't want to make the example JSON too big.

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tratcher) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:372]: if there's _no_ matching pattern?

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 3]
```diff
@@ -314,7 +321,59 @@ webBuilder.ConfigureKestrel(serverOptions =>
...
+        "Url": "https://*",
+        "Sni": {
+          "a.example.org": {
+            "Protocols": "Http1AndHttp2",
+            "SslProtocols": [ "Tls11", "Tls12", "Tls13"],
+            "Certificate": {
+              "Subject": "<subject; required>",
+              "Store": "<certificate store; required>",
+            },
+            "ClientCertificateMode" : "NoCertificate"
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel/endpoints.md:342]: The code-first approach relies on API docs to explain what parameters and values can be used. Do we need to provide more explicit docs here for the config approach so people will understand the parameters and values? The sample is pretty good, but it's not clear if it's exhaustive. Or are these all documented elsewhere that we can link to?

Certs: #certificate-sources
Protocols: #set-the-http-protocol-from-configuration
SslProtocols: https://docs.microsoft.com/en-us/dotnet/api/system.security.authentication.sslprotocols?view=net-5.0 (and expand #ssltls-protocols)
ClientCertificateMode: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.https.clientcertificatemode?view=aspnetcore-5.0

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (halter73) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:372]: > if there's _no_ matching pattern?

Yes. That was a rather important word to omit 😆 

> The fallback flow for endpoint configuration is documented elseware. I don't think it needs to be repeated.

Sounds good.

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (halter73) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:355]: I completely understand not wanting to make the example JSON to large, but I think it's worth adding a line for this. I think SNI configuration is the first time config can ever _fall back_ to the configuration at the root of the endpoint config so I think we should make it very clear that can happen.

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md (hunk 3 above)
Comment (halter73) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:342]: Good point. This is exhaustive, but it's not clear that's the case without saying so explicitely somewhere.

**Comment by JamesNK** [comment]
All feedback applied.

**Review by Tratcher — APPROVED** [review]

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 4]
```diff
@@ -272,6 +272,13 @@ webBuilder.ConfigureKestrel(serverOptions =>
 
 [Server Name Indication (SNI)](https://tools.ietf.org/html/rfc6066#section-3) can be used to host multiple domains on the same IP address and port. For SNI to function, the client sends the host name for the secure session to the server during the TLS handshake so that the server can provide the correct certificate. The client uses the furnished certificate for encrypted communication with the server during the secure session that follows the TLS handshake.
 
+SNI can be configured in two ways:
+
+* Create an endpoint in code and select a certificate using the host name with the `ServerCertificateSelector` callback.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/endpoints.md:277]:
```suggestion
* Create an endpoint in code and select a certificate using the host name with the <xref:Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions.ServerCertificateSelector%2A> callback.
```

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 5]
```diff
@@ -272,6 +272,13 @@ webBuilder.ConfigureKestrel(serverOptions =>
...
+* Configure a mapping between host names and HTTPS options in configuration. For example, JSON in `appsettings.json`.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/servers/kestrel/endpoints.md:278]:
```suggestion
* Configure a mapping between host names and HTTPS options in [Configuration](xref:fundamentals/configuration/index). For example, JSON in  the `appsettings.json` file.
```

**Review by halter73 — APPROVED** [review]

**Review by Rick-Anderson — APPROVED** [review]
Suggestions, feel free to ignore. I think I have the xref correct. 
`<xref:Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions.ServerCertificateSelector%2A`

**Pull request merged**